python/cuml/ensemble/randomforest_shared.pyx

```diff
@@ -0,0 +1,122 @@
+from cython.operator cimport dereference as deref, preincrement as inc
+from cpython.object cimport PyObject
+from libc.stdint cimport uintptr_t
+from typing import Tuple, Dict, List, Union
+import numpy as np
+
+cdef extern from "treelite/tree.h" namespace "treelite":
+    cdef struct PyBufferFrame:
+        void* buf
+        char* format
+        size_t itemsize
+        size_t nitem
+    cdef cppclass Model:
+        vector[PyBufferFrame] GetPyBuffer() except +
+        void InitFromPyBuffer(vector[PyBufferFrame] frames) except +
+
+cdef extern from "Python.h":
+    Py_buffer* PyMemoryView_GET_BUFFER(PyObject* mview)
+
+cdef class PyBufferFrameWrapper:
+    cdef PyBufferFrame _handle
+    cdef Py_ssize_t shape[1]
+    cdef Py_ssize_t strides[1]
+
+    def __cinit__(self):
+        pass
+
+    def __dealloc__(self):
+        pass
+
+    def __getbuffer__(self, Py_buffer* buffer, int flags):
+        cdef Py_ssize_t itemsize = self._handle.itemsize
+
+        self.shape[0] = self._handle.nitem
+        self.strides[0] = itemsize
+
+        buffer.buf = self._handle.buf
+        buffer.format = self._handle.format
+        buffer.internal = NULL
+        buffer.itemsize = itemsize
+        buffer.len = self._handle.nitem * itemsize
+        buffer.ndim = 1
+        buffer.obj = self
+        buffer.readonly = 0
+        buffer.shape = self.shape
+        buffer.strides = self.strides
+        buffer.suboffsets = NULL
+
+    def __releasebuffer__(self, Py_buffer *buffer):
+        pass
+
+cdef PyBufferFrameWrapper MakePyBufferFrameWrapper(PyBufferFrame handle):
+    cdef PyBufferFrameWrapper wrapper = PyBufferFrameWrapper()
+    wrapper._handle = handle
+    return wrapper
+
+cdef list _get_frames(ModelHandle model):
+    return [memoryview(MakePyBufferFrameWrapper(v))
+            for v in (<Model*>model).GetPyBuffer()]
+
+cdef ModelHandle _init_from_frames(vector[PyBufferFrame] frames) except *:
+    cdef Model* model_obj = new Model()
+    model_obj.InitFromPyBuffer(frames)
+    return <ModelHandle>model_obj
+
+
+def get_frames(model: uintptr_t) -> List[memoryview]:
+    return _get_frames(<ModelHandle> model)
+
+
+def init_from_frames(frames: List[np.ndarray],
+                     format_str: List[str], itemsize: List[int]) -> uintptr_t:
+    cdef vector[PyBufferFrame] cpp_frames
+    cdef Py_buffer* buf
+    cdef PyBufferFrame cpp_frame
+    for i, frame in enumerate(frames):
+        x = memoryview(frame)
+        buf = PyMemoryView_GET_BUFFER(<PyObject*>x)
+        cpp_frame.buf = buf.buf
+        cpp_frame.format = format_str[i]
+        cpp_frame.itemsize = itemsize[i]
+        cpp_frame.nitem = buf.len // itemsize[i]
+        cpp_frames.push_back(cpp_frame)
+    return <uintptr_t> _init_from_frames(cpp_frames)
+
+
+def treelite_serialize(
+    model: uintptr_t
+) -> Dict[str, Union[List[str], List[np.ndarray]]]:
+    frames = get_frames(model)
+    header = {'format_str': [x.format.encode('utf-8') for x in frames],
```

Comment:
```suggestion
    header = {'format_str': [x.format for x in frames],
```

Reply:
Is it preferable to pass `str` to pickle, rather than `bytes`?

Reply:
Was thinking about this in the context of moving from pickle to Dask serialization down the road (assuming that is still the plan). Typically the header consists of things like `int`, `str`, `dict`, `list`, and `tuple`. Generally things that are MsgPack serializable. Typically `bytes` and `memoryviews` are reserved for frames instead.

Was unsure at first whether `bytes` would work in the header. However after playing with things a bit `bytes` may work. MsgPack is at least able to handle them with the flags that Dask is using.

Reply:
Good point. Thanks for your explanation. I agree that built-in types like `str` would be well supported by Dask serializer.